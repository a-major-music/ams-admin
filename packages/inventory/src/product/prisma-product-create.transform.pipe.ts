import { PipeTransform, Injectable, ArgumentMetadata } from '@nestjs/common';

import * as _ from 'lodash';

/* 
The types defined in @amm/types are related, and these relationships mirror the structure of the database. However,
Prisma does not use DTOs and repositories like other ORMs. Instead it uses a JSON format to define both data and
operation in the same structure. Top level items are the same, but nested objects need operations inserted in the
object tree to tell Prisma what to do. So a create({id: 1, nested: {id: 1}}) becomes 
create({id: 1, nested: create: {id: 1}}). This interceptor inserts these operations into the objects before they hit
the controller to create variants, and to connectOrCreate suppliers and tags (to avoid duplicates)
*/
@Injectable()
export class PrismaProductCreateTransformPipe implements PipeTransform {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  transform(value: any, metadata: ArgumentMetadata) {
    const augments: any = {};

    // True if the product has more than 1 variant
    const hasVariants = value.variants.filter(v => !v.pendingDelete).length > 1;

    if (!_.isEmpty(value.suppliers)) {
      augments.suppliers = {
        connect: value.suppliers.map((s) => {
          return { guid: s.guid };
        }),
      };
    } else {
      value.suppliers = undefined;
    }

    if (!_.isEmpty(value.variants)) {
      augments.variants = { create: [] };

      // If there is only one variant we don't store filterValues and filterFields. You don't need them if
      // there is a single variant, and it improves the look of the storefront.
      value.variants.forEach((variant) => {
        if (hasVariants) {
          variant.filterValues = variant.filterValues.toString();
        }
        else {
          variant.filterValues = "";
        }

        if (!_.isEmpty(variant.imageUrls)) {
          const newField = { create: [] };
          variant.imageUrls.forEach((i) => {
            newField.create.push({ uri: i });
          });
          variant.imageUrls = newField;
        } else {
          variant.imageUrls = undefined;
        }

        augments.variants.create.push(variant);
      });
    }

    if (hasVariants) {
      augments.filterFields = value.filterFields.toString();
    }
    else {
      augments.filterFields = "";
    }

    return { ...value, ...augments };
  }
}
