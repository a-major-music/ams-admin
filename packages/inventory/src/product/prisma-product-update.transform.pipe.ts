import { PipeTransform, Injectable, ArgumentMetadata, Logger } from '@nestjs/common';

import * as _ from 'lodash';

/* 
The types defined in @amm/types are related, and these relationships mirror the structure of the database. However,
Prisma does not use DTOs and repositories like other ORMs. Instead it uses a JSON format to define both data and
operation in the same structure. Top level items are the same, but nested objects need operations inserted in the
object tree to tell Prisma what to do. So a create({id: 1, nested: {id: 1}}) becomes 
create({id: 1, nested: create: {id: 1}}). This interceptor inserts these operations into the objects before they hit
the controller to create variants, and to connectOrCreate suppliers and tags (to avoid duplicates)
*/
@Injectable()
export class PrismaProductUpdateTransformPipe implements PipeTransform {
  logger = new Logger(PrismaProductUpdateTransformPipe.name);
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  transform(value: any, metadata: ArgumentMetadata) {
    const augments: any = {};

    // True if the product has more than 1 variant
    const hasVariants =
      value.variants.filter((v) => !v.pendingDelete).length > 1;

    if (!_.isEmpty(value.suppliers)) {
      augments.suppliers = {
        // Delete existing suppier - DB supports many-to-many, but not ready to support that yet
        deleteMany: {},
        connect: value.suppliers.map((s) => {
          return { guid: s.guid };
        }),
      };
    } else {
      value.suppliers = undefined;
    }

    if (!_.isEmpty(value.variants)) {
      augments.variants = { upsert: [] };

      value.variants.forEach((variant) => {
        if (variant.pendingDelete === true) {
          if (!augments.variants.delete) {
            augments.variants.delete = [];
          }

          augments.variants.delete.push({ id: variant.id });
        } else {
          augments.variants.pendingDelete = undefined;

          // Remove stock levels data from the variants - this should only be affected by explicit stock level
          // update calls to prevent accidental overwriting of stock values
          // delete variant.stockOnHand;

          if (hasVariants) {
            variant.filterValues = variant.filterValues.toString();
          } else {
            variant.filterValues = '';
          }

          if (!_.isEmpty(variant.imageUrls)) {
            const imageUrlField = { connectOrCreate: [] };
            variant.imageUrls.forEach((i) => {
              if (i.guid) {
                imageUrlField.connectOrCreate.push({
                  where: { guid: i.guid },
                  create: i,
                });
              } else {
                imageUrlField.connectOrCreate.push({
                  where: { uri: i },
                  create: { uri: i },
                });
              }
            });
            variant.imageUrls = imageUrlField;
          } else {
            variant.imageUrls = undefined;
          }

          const where: any = variant.guid
            ? { guid: variant.guid }
            : { sku: variant.sku };

          augments.variants.upsert.push({
            // Remove stock levels from and update, and IDs from a create call
            where,
            update: { ...variant, stockOnHand: undefined },
            create: { ...variant, id: undefined },
          });
        }
      });
    }

    if (hasVariants) {
      augments.filterFields = value.filterFields.toString();
    } else {
      augments.filterFields = '';
    }

    return { ...value, ...augments };
  }
}
