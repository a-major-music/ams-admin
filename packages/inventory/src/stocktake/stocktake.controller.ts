import {
  Body,
  Controller,
  Delete,
  Get,
  HttpException,
  Param,
  Post,
  Put,
} from '@nestjs/common';
import { StocktakeLine } from '@amm/types';
import { StocktakeService } from './stocktake.service';

@Controller('stocktake')
export class StocktakeController {
  constructor(private stocktakeService: StocktakeService) {}

  @Post('save/:location')
  save(@Param('location') location: string, @Body() lines: StocktakeLine[]) {
    if (!location || location === '') {
      throw new HttpException(
        'Please provide a location for the stock take',
        400,
      );
    }

    return this.stocktakeService.save(location, lines);
  }

  @Get('findByLocation/:location')
  findByLocation(@Param('location') location: string) {
    if (!location || location === '') {
      throw new HttpException('Please provide a location to find', 400);
    }

    return this.stocktakeService.findByLocation(location);
  }

  @Get()
  findAll() {
    return this.stocktakeService.findAll();
  }

  @Put('apply/:location')
  apply(@Param('location') location) {
    if (!location || location === '') {
      throw new HttpException('Please provide a location to apply', 400);
    }

    return this.stocktakeService.apply(location);
  }

  @Delete('delete/:location')
  delete(@Param('location') location: string) {
    if (!location || location === '') {
      throw new HttpException(
        'Please provide a location for the stock take',
        400,
      );
    }
    return this.stocktakeService.deleteByLocation(location);
  }
}
