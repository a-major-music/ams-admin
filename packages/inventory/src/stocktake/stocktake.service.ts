import { Injectable } from '@nestjs/common';
import { StocktakeLine } from '@amm/types';
import { PrismaService } from '../prisma.service';

@Injectable()
export class StocktakeService {
  constructor(private prisma: PrismaService) {}

  async findAll() {
    return this.prisma.stocktake.findMany({ include: { lines: true } });
  }
  async findByLocation(location: string) {
    return this.prisma.stocktake.findFirst({
      where: { location: location },
      include: {
        lines: { include: { variant: { include: { Product: true } } } },
      },
    });
  }

  async deleteByLocation(location: string) {
    return this.prisma.stocktake.deleteMany({ where: { location: location } });
  }

  async save(location: string, lines: StocktakeLine[]) {
    const formattedLines: StocktakeLine[] = lines.map((line) => {
      return {
        count: line.count,
        productVariantId: line.variant?.id,
      };
    });
    return this.prisma.stocktake.upsert({
      where: { location: location },
      update: {
        lines: {
          deleteMany: {},
          createMany: { data: formattedLines },
        },
      },
      create: { location: location, lines: { create: formattedLines } },
      include: { lines: { include: { variant: true } } },
    });
  }

  async apply(location: string) {
    return this.prisma.stocktake.update({
      where: { location },
      data: { applied: true },
    });
  }
}
