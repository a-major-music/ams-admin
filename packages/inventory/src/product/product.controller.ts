import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  Query,
  HttpStatus,
  HttpException,
  Logger,
} from '@nestjs/common';

import { ProductService } from './product.service';

import { Product } from '@amm/types';
import { ApiOperation, ApiParam, ApiTags } from '@nestjs/swagger';
import { PrismaProductCreateTransformPipe } from './prisma-product-create.transform.pipe';
import { ProductVariant } from '../../prisma/generated/client';

import _ from 'lodash';
import { PrismaProductUpdateTransformPipe } from './prisma-product-update.transform.pipe';
import { ShopifyService } from '../shopify/shopify.service';

@ApiTags('products')
@Controller('products')
export class ProductController {
  constructor(
    private productService: ProductService,
    private shopifyService: ShopifyService,
  ) {}

  private readonly logger = new Logger(ProductController.name);

  @Get()
  @ApiOperation({
    description:
      'Returns all products as array of Product, with deep fetch of nested objects',
  })
  @ApiParam({
    name: 'skip',
    required: false,
    description: 'number of records to skip',
    type: 'integer',
  })
  @ApiParam({
    name: 'take',
    required: false,
    description: 'number of records to return',
    type: 'integer',
  })
  async findAll(
    @Query('skip') skip?: number,
    @Query('take') take?: number,
    @Query('search') search?: string,
  ): Promise<Product[]> {
    const params: any = {};
    if (skip) params.skip = +skip;
    if (take) {
      params.take = +take;
    } else {
      params.take = 100;
    }
    if (search) {
      // convert to a Prisma where clause to check the name and SKU for the search
      params.where = {
        OR: [
          { name: { contains: search, mode: 'insensitive' } },
          {
            variants: {
              some: {
                OR: [
                  { sku: { contains: search, mode: 'insensitive' } },
                  { barcode: { contains: search } },
                  { filterValues: { contains: search, mode: 'insensitive' } },
                ],
              },
            },
          },
        ],
      };
    }

    return this.productService.findMany(params) as any;
  }

  @Get('forLocation/:location')
  async findByLocation(@Param('location') location) {
    return this.productService.findMany({
      where: { location: location },
      take: 1000,
    });
  }

  @Get('count')
  async countProducts(): Promise<any> {
    return { rows: (await this.productService.countProducts())._count.id };
  }

  @Get('variantBySku/:sku')
  async findByVariantSku(
    @Param('sku') sku: string,
    @Query('variantOnly') variantOnly = false,
  ): Promise<Product[] | ProductVariant> {
    return this.productService.findByVariantSku(sku, variantOnly) as any;
  }

  @Post('variantsForSkus')
  async variantsForSkus(
    @Body('skus') skus: string[],
  ): Promise<ProductVariant[]> {
    if (!skus || skus.length == 0) return null;
    return this.productService.variantsForSkus(skus) as any;
  }

  @Get('variant/:guid')
  async findByVariant(
    @Param('guid') guid: string,
  ): Promise<Product[] | ProductVariant> {
    return this.productService.findByVariantGuid(guid) as any;
  }

  @Post('inventory')
  async getInventoryMap(
    @Body('skus') skus: string[],
    @Body('guids') guids: string[],
  ) {
    if (!skus && !guids) {
      throw new HttpException(
        'Please provide either a list of SKUs or GUIDs',
        HttpStatus.BAD_REQUEST,
      );
    }

    if (skus) {
      return this.productService.getSkuToInventoryMap(skus);
    } else if (guids) {
      return this.productService.getGuidToInventoryMap(guids);
    }
  }

  @Get('variantByBarcode/:barcode')
  async findVariantByBarcode(
    @Param('barcode') barcode: string,
  ): Promise<ProductVariant | null> {
    return this.productService.findVariantByBarcode(barcode);
  }

  @Get('variants-guid-map')
  async getSkuToGuidMap(): Promise<any[]> {
    return this.productService.getSkuToGuidMap();
  }

  @Get('stockValue')
  async getStockValue(@Query('date') date: string): Promise<any> {
    return await this.productService.getStockValue(date);
  }

  // Make sure this is the last get, otherwise it will grab all GET requests.
  @Get(':guid')
  async findOne(@Param('guid') guid: string): Promise<Product> {
    return this.productService.findByGuid(guid).then(async (p: Product) => {
      if (p && p.publishedToShopify) {
        const shopifyId = await this.shopifyService.findProductIdByVariantSkus(
          p.variants.map((v) => v.sku),
        );

        if (shopifyId) {
          p.shopifyId = shopifyId;
        }
      }

      return p;
    });
  }

  @Post('create')
  async create(@Body(PrismaProductCreateTransformPipe) p: Product) {
    return await this.productService.create(p as any).catch((e) => {
      if (e.code === 'P2002') {
        const message = `The ${e.meta.target[0]} field has already been used, but must be unique!`;
        throw new HttpException(message, HttpStatus.CONFLICT);
      }

      this.logger.error(e);
      throw new HttpException(e, HttpStatus.BAD_REQUEST);
    });
  }

  @Put('/update-inventory')
  async updateInventory(
    @Body()
    inventory: {
      variantSku: string; // Shopify is keyed on SKU, so need this in addition to the GUID
      variantGuid: string;
      stockOnHand: number;
    }[],
  ) {
    // There is no await here, as the shopify part may take a while so we return.
    return this.productService
      .updateInventory(inventory)
      .then(() => this.shopifyService.updateInventory(inventory));
  }

  @Put(':sku/updateMedia/:mediaSrc')
  async updateMedia(
    @Param('sku') sku: string,
    @Param('mediaSrc') mediaSrc: string,
  ) {
    return await this.productService.updateVariantMedia(sku, mediaSrc);
  }

  @Put(':guid/update')
  async update(
    // This flag prevents updates FROM Shopify immediately causing a pointless sync to Shopify
    @Query('updateShopify') updateShopify: string,
    @Param('guid') guid: string,
    @Body(PrismaProductUpdateTransformPipe) productData: Product,
  ) {
    productData.guid = guid;
    const params = { where: { guid: guid }, data: productData };

    return await this.productService
      .update(params as any)
      .then(async (p) => {
        // if this is published to Shopify try to update that too, synchronously, so we can tell the
        // user if it didn't work.
        if (p.publishedToShopify && updateShopify === 'true') {
          this.logger.log(
            'Product is flagged as published, so attempting Shopify update',
          );
          const res = await this.shopifyService.updateProduct(p);
          const shopifyId = res.productSet?.product?.id;

          if (!shopifyId) {
            this.logger.warn(
              `Cannot retrieve product ID following create / update of [${p.name}] - reversing publishedToShopify flag`,
            );
            return this.productService.update({
              where: { id: p.id },
              data: { shopifyId: '', publishedToShopify: false },
            });
          }

          this.logger.debug(
            `Updating product [${p.name}] with shopify ID [${shopifyId}]`,
          );

          return this.productService.update({
            where: { id: p.id },
            data: { shopifyId: shopifyId },
          });
        } else {
          // Delete this from Shopify - may not be there, but just in case the product was unpublished
          await this.shopifyService.deleteProduct(p.name);

          return this.productService.update({
            where: { id: p.id },
            data: { shopifyId: '', publishedToShopify: false },
          });
        }
      })
      .catch((e) => {
        if (e.code === 'P2002') {
          const message = `The ${e.meta.target[0]} field has already been used, but must be unique!`;
          throw new HttpException(message, HttpStatus.CONFLICT);
        }

        this.logger.error(e);
        throw new HttpException(e, HttpStatus.BAD_REQUEST);
      });
  }

  @Put(':guid/update-stock')
  async updateStock(
    // This flag prevents updates FROM Shopify immediately causing a pointless sync to Shopify
    @Query('updateShopify') updateShopify: string,
    @Param('guid') guid: string,
    @Body() body: any,
  ): Promise<any> {
    if (!guid || guid.trim() === '') {
      throw new HttpException(
        'Please provide a variant GUID',
        HttpStatus.BAD_REQUEST,
      );
    }

    const stockLevelDelta = +body.stockLevelDelta || undefined;
    const newBuyPrice = +body.newBuyPrice || undefined;
    const newRetailPrice = +body.newRetailPrice || undefined;

    if (
      !newBuyPrice &&
      !newRetailPrice &&
      (!stockLevelDelta || stockLevelDelta == 0)
    ) {
      throw new HttpException(
        'Please provide something to update',
        HttpStatus.BAD_REQUEST,
      );
    }

    this.productService
      .updateStock({
        variantGuid: guid,
        stockLevelDelta: stockLevelDelta,
        newBuyPrice: newBuyPrice,
        newRetailPrice: newRetailPrice,
      })
      .then(async (v) => {
        if (updateShopify === 'true') {
          const p: Product[] = (await this.productService.findByVariantGuid(
            guid,
          )) as Product[];

          this.logger.log(
            `Syncing product prices and stock level with Shopify for ${v.sku}`,
          );

          if (!p || p.length == 0) {
            this.logger.error(
              `Unable to update product with variant ${guid} following receipt - could not find product!`,
            );
          } else {
            // Sync the product with shopify to put the new prices
            this.shopifyService.updateProduct(p[0]);
          }

          // Stock levels are explicitly not updated with the previous call, so do that too.
          this.shopifyService.updateInventory([
            {
              variantSku: v.sku,
              stockOnHand: v.stockOnHand,
            },
          ]);
        }
      })
      .catch((e) => {
        this.logger.error(e);
        throw new HttpException(e.toString(), HttpStatus.BAD_REQUEST);
      });
  }

  @Put(':productGuid/update-location/:location')
  async updateLocation(
    @Param('productGuid') productGuid,
    @Param('location') location,
  ) {
    return await this.productService.update({
      where: { guid: productGuid },
      data: { location: location },
    });
  }

  @Put('/sync-stock-levels')
  async syncStockLevels() {
    // Calls updateInventory page by page for the whole published catalogue.
    // This is intended to be run overnight. Returns 202 - ACCEPTED and processes in the background
    new Promise(async (resolve) => {
      let results: Product[] = [];
      let cursor = 0;
      let totalProcessed = 0;

      while (true) {
        results = await this.productService.findMany({
          take: 50,
          cursor: cursor > 0 ? { id: cursor } : undefined,
          skip: cursor == 0 ? 0 : 1, // Skip the cursor, unless it's the first time through
          where: { publishedToShopify: true },
          orderBy: { id: 'asc' },
        });

        if (results.length == 0) {
          // No more results, so break out of this loop
          break;
        }

        cursor = results.slice(-1)[0].id;

        const variants: { variantSku: string; stockOnHand: number }[] = results
          .map((p) => p.variants)
          .flat()
          .map((pv) => {
            return { variantSku: pv.sku, stockOnHand: pv.stockOnHand };
          });

        totalProcessed += variants.length;

        // Update stock levels in Shopify for these variants
        await this.shopifyService.updateInventory(variants);
      }

      this.logger.log(`Processed ${totalProcessed} variants`);

      resolve({ status: 'success', processed: totalProcessed });
    });

    this.logger.log(
      'Request to sync stock accepted, and in processing. Check console for updates.',
    );
    return HttpStatus.ACCEPTED;
  }

  @Delete(':guid')
  delete(@Param('guid') guid: string) {
    this.productService
      .delete(guid)
      .then((p: Product) => {
        return this.shopifyService.deleteProduct(p.name);
      })
      .catch((e) => {
        this.logger.error(e);
        throw new HttpException(e.toString(), HttpStatus.BAD_REQUEST);
      });
  }

  @Delete('/')
  async deleteMany(@Query('where') where: any): Promise<any> {
    return this.productService.deleteMany(where).then((deleted) => {
      return Promise.all(
        deleted.forEach(
          async (p) => await this.shopifyService.deleteProduct(p.name),
        ),
      );
    });
  }

  @Put('recordStockValue')
  recordStockValue() {
    return this.productService.recordStockValue();
  }
}
