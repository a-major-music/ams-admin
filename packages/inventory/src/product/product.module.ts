import { Module } from '@nestjs/common';
import { PrismaService } from '../prisma.service';
import { ProductService } from './product.service';
import { ProductController } from './product.controller';
import { HttpModule } from '@nestjs/axios';
import { ShopifyService } from '../shopify/shopify.service';

@Module({
  imports: [HttpModule],
  providers: [ProductService, ShopifyService, PrismaService],
  controllers: [ProductController],
})
export class ProductModule {}
