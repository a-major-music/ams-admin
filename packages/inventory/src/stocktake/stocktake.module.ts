import { Module } from '@nestjs/common';
import { PrismaService } from '../prisma.service';
import { StocktakeController } from './stocktake.controller';
import { StocktakeService } from './stocktake.service';

@Module({
  controllers: [StocktakeController],
  providers: [StocktakeService, PrismaService],
})
export class StocktakeModule {}
