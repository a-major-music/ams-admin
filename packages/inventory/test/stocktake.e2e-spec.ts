import { INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { Stocktake } from '@amm/types';
import supertest = require('supertest');
import { AppModule } from '../src/app.module';
import { PrismaService } from '../src/prisma.service';
import { removeProps } from './util';

import { Product, ProductType } from '@amm/types';
import { ProductModule } from '../src/product/product.module';
import { StocktakeModule } from '../src/stocktake/stocktake.module';

let app: INestApplication;
let prisma: PrismaService;
let testProduct: Product;

beforeAll(async () => {
  const module = await Test.createTestingModule({
    imports: [AppModule, StocktakeModule, ProductModule],
  }).compile();

  app = module.createNestApplication();
  prisma = app.get<PrismaService>(PrismaService);
  await app.init();

  testProduct = _createTestProduct();

  console.log(testProduct);
});

afterAll(async () => {
  prisma.product.deleteMany({});
  await app.close();
});

const _createTestProduct = () => {
  return prisma.product.create({
    data: {
      name: 'Testing Stocktake',
      taxable: true,
      productType: ProductType.INSTRUMENT,
      brand: 'Test Brand',
      description: 'Test Description',
      variants: {
        create: [
          {
            filterValues: 'small',
            notes: 'A small test variant',
            retailPrice: 1500,
            sku: 'INS00001-1',
            stockOnHand: 0,
            weight: 1,
            weightUnit: 'Kg',
            manageStockLevels: true,
            buyPrice: 750,
          },
          {
            filterValues: 'medium',
            notes: 'A medium test variant',
            retailPrice: 1500,
            sku: 'INS00001-2',
            stockOnHand: 0,
            weight: 1,
            weightUnit: 'Kg',
            manageStockLevels: true,
            buyPrice: 750,
          },
          {
            filterValues: 'large',
            notes: 'A large test variant',
            retailPrice: 1500,
            sku: 'INS00001-3',
            stockOnHand: 0,
            weight: 1,
            weightUnit: 'Kg',
            manageStockLevels: true,
            buyPrice: 750,
          },
        ],
      },
      filterFields: 'size',
      location: 'A1',
    },
  });
};

describe('POST /stocktake/save', () => {
  afterAll(async () => {
    await prisma.stocktake.deleteMany({});
  });

  it('should save a minimal stocktake', async () => {
    const s: Stocktake = {
      location: 'TEST',
      lines: [
        {
          count: 1,
          variant: testProduct.variants[0],
        },
      ],
    };

    await supertest
      .agent(app.getHttpServer())
      .post(`/stocktake/save/${s.location}`)
      .set('Accept', 'application/json')
      .send(s.lines)
      .expect('Content-Type', /json/)
      .expect(201);

    const qs = await supertest
      .agent(app.getHttpServer())
      .get(`/stocktake/findByLocation/${s.location}`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(JSON.parse(qs.text), ['id', 'guid'])).toMatchSnapshot();
  });
});

describe.skip('GET /stocktake', () => {
  const mocks: Stocktake[] = [
    {
      location: 'LOC1',
      lines: [
        {
          variant: testProduct.variants[0],
          count: 1,
        },
      ],
    },
    {
      location: 'LOC2',
      lines: [
        {
          variant: testProduct.variants[1],
          count: 1,
        },
      ],
    },
  ];

  beforeAll(async () => {
    await Promise.all(
      mocks.map((mock) => {
        return supertest
          .agent(app.getHttpServer())
          .post(`/stocktake/save/${mock.location}`)
          .set('Accept', 'application/json')
          .send(mock.lines)
          .expect('Content-Type', /json/)
          .expect(201);
      }),
    );
  });

  afterAll(async () => {
    await prisma.stocktake.deleteMany({});
  });

  it('should return all stocktakes', async () => {
    const qs = await supertest
      .agent(app.getHttpServer())
      .get(`/stocktake/`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(JSON.parse(qs.text), ['id', 'guid'])).toMatchSnapshot();
  });

  it('should return one stocktake', async () => {
    const qs = await supertest
      .agent(app.getHttpServer())
      .get(`/stocktake/findByLocation/${mocks[1].location}`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(JSON.parse(qs.text), ['id', 'guid'])).toMatchSnapshot();
  });
});

describe.skip('POST /stocktake/save', () => {
  const mock = {
    location: 'LOC1',
    lines: [
      {
        variant: testProduct.variants[0],
        count: 1,
      },
    ],
  };

  beforeAll(async () => {
    await supertest
      .agent(app.getHttpServer())
      .post(`/stocktake/save/${mock.location}`)
      .set('Accept', 'application/json')
      .send(mock.lines)
      .expect('Content-Type', /json/)
      .expect(201);
  });

  afterAll(async () => {
    await prisma.stocktake.deleteMany({});
  });

  it('should update count of INS0001', async () => {
    mock.lines[0].count += 1;
    const qs = await supertest
      .agent(app.getHttpServer())
      .post(`/stocktake/save/${mock.location}`)
      .set('Accept', 'application/json')
      .send(mock.lines)
      .expect('Content-Type', /json/)
      .expect(201);

    expect(removeProps(JSON.parse(qs.text), ['id', 'guid'])).toMatchSnapshot();
  });

  it('should add new count for INS0002', async () => {
    mock.lines.push({ variant: { id: 2 }, count: 2 });
    const qs = await supertest
      .agent(app.getHttpServer())
      .post(`/stocktake/save/${mock.location}`)
      .set('Accept', 'application/json')
      .send(mock.lines)
      .expect('Content-Type', /json/)
      .expect(201);

    expect(removeProps(JSON.parse(qs.text), ['id', 'guid'])).toMatchSnapshot();
  });

  it('should add new count for INS0003', async () => {
    mock.lines.push({ variant: testProduct.variants[0], count: 3 });
    const qs = await supertest
      .agent(app.getHttpServer())
      .post(`/stocktake/save/${mock.location}`)
      .set('Accept', 'application/json')
      .send(mock.lines)
      .expect('Content-Type', /json/)
      .expect(201);

    expect(removeProps(JSON.parse(qs.text), ['id', 'guid'])).toMatchSnapshot();
  });
});

describe.skip('DELETE /stocktake', () => {
  const mocks = [
    {
      location: 'LOC1',
      lines: [
        {
          variant: testProduct.variants[0],
          count: 1,
        },
      ],
    },
    {
      location: 'LOC2',
      lines: [
        {
          variant: testProduct.variants[1],
          count: 2,
        },
      ],
    },
  ];

  beforeAll(async () => {
    await Promise.all(
      mocks.map((mock) => {
        return supertest
          .agent(app.getHttpServer())
          .post(`/stocktake/save/${mock.location}`)
          .set('Accept', 'application/json')
          .send(mock.lines)
          .expect('Content-Type', /json/)
          .expect(201);
      }),
    );
  });

  afterAll(async () => {
    await prisma.stocktake.deleteMany({});
  });

  it('should delete the first stock take only', async () => {
    await supertest
      .agent(app.getHttpServer())
      .delete(`/stocktake/delete/${mocks[0].location}`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    const qs = await supertest
      .agent(app.getHttpServer())
      .get(`/stocktake/`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(JSON.parse(qs.text), ['id', 'guid'])).toMatchSnapshot();
  });
});
