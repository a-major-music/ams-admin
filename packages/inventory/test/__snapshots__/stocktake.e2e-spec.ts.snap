// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DELETE /stocktake should delete the first stock take only 1`] = `
Array [
  Object {
    "lines": Array [
      Object {
        "count": 2,
        "sku": "INS0002",
        "stockTakeId": 6,
      },
    ],
    "location": "LOC2",
  },
]
`;

exports[`GET /stocktake should return all stocktakes 1`] = `
Array [
  Object {
    "lines": Array [
      Object {
        "count": 1,
        "sku": "INS0001",
        "stockTakeId": 2,
      },
    ],
    "location": "LOC1",
  },
  Object {
    "lines": Array [
      Object {
        "count": 1,
        "sku": "INS0002",
        "stockTakeId": 3,
      },
    ],
    "location": "LOC2",
  },
]
`;

exports[`GET /stocktake should return one stocktake 1`] = `
Object {
  "lines": Array [
    Object {
      "count": 1,
      "sku": "INS0002",
      "stockTakeId": 3,
    },
  ],
  "location": "LOC2",
}
`;

exports[`POST /stocktake/save should add new count for INS0002 1`] = `
Object {
  "lines": Array [
    Object {
      "count": 2,
      "sku": "INS0001",
      "stockTakeId": 4,
    },
    Object {
      "count": 2,
      "sku": "INS0002",
      "stockTakeId": 4,
    },
  ],
  "location": "LOC1",
}
`;

exports[`POST /stocktake/save should add new count for INS0003 1`] = `
Object {
  "lines": Array [
    Object {
      "count": 2,
      "sku": "INS0001",
      "stockTakeId": 4,
    },
    Object {
      "count": 2,
      "sku": "INS0002",
      "stockTakeId": 4,
    },
    Object {
      "count": 3,
      "sku": "INS0003",
      "stockTakeId": 4,
    },
  ],
  "location": "LOC1",
}
`;

exports[`POST /stocktake/save should save a minimal stocktake 1`] = `
Object {
  "lines": Array [
    Object {
      "count": 1,
      "sku": "INS0001",
      "stockTakeId": 1,
    },
  ],
  "location": "TEST",
}
`;

exports[`POST /stocktake/save should update count of INS0001 1`] = `
Object {
  "lines": Array [
    Object {
      "count": 2,
      "sku": "INS0001",
      "stockTakeId": 4,
    },
  ],
  "location": "LOC1",
}
`;
